examples/shor/functions.py

```diff
@@ -0,0 +1,420 @@
+    else:
+        g, y, x = egcd(b % a, a)
+        return (g, x - (b // a) * y, y)
+
+
+def mod_inv(a, N):
+    """Classical algorithm to calculate the inverse of a modulo N.
+    Args:
+        a (int): number to invert.
+        N (int): modulo.
+
+    Returns:
+        x%N (int): inverse of a modulo N.
+    """
+    g, x, y = egcd(a, N)
+    if g != 1:
+        raise ValueError('modular inverse does not exist')
+    else:
+        return x % N
+
+
+def quantum_order_finding_full(N, a):
+    """Quantum circuit that performs the order finding algorithm using a fully quantum iQFT.
+    Args:
+        N (int): number to factorize.
+        a (int): chosen number to use in the algorithm.
+
+    Returns:
+        s (float): value of the state measured by the quantum computer.
+    """
+    print('  - Performing algorithm using a fully quantum iQFT.\n')
+    # Creating the parts of the needed quantum circuit.
+    n = int(np.ceil(np.log2(N)))
+    b = [i for i in range(n+1)]
+    x = [n+1+i for i in range(n)]
+    ancilla = 2*n + 1
+    q_reg = [2*n + 2 + i for i in range(2*n)]
+    circuit = Circuit(4*n+2)
+    print(f'  - Total number of qubits used: {4*n+2}.\n')
+
+    # Building the quantum circuit
+    for i in range(len(q_reg)):
+        circuit.add(gates.H(q_reg[i]))
+    circuit.add(gates.X(x[len(x)-1]))
+    circuit.add((c_U(q, x, b, a**(2**i), N, ancilla, n) for i, q in enumerate(q_reg)))
+    circuit.add(i_qft(q_reg))
+
+    # Adding measurement gates
+    circuit.add(gates.M(*q_reg))
+    result = circuit(nshots=1)
+    s = result.frequencies(binary=False).most_common()[0][0]
+    print(f"The quantum circuit measures s = {s}.\n")
+    return s
+
+
+def quantum_order_finding_semiclassical(N, a):
+    """Quantum circuit that performs the order finding algorithm using a semiclassical iQFT.
+    Args:
+        N (int): number to factorize.
+        a (int): chosen number to use in the algorithm.
+
+    Returns:
+        s (float): value of the state measured by the quantum computer.
+    """
+    print('  - Performing algorithm using a semiclassical iQFT.\n')
+    # Creating the parts of the needed quantum circuit.
+    n = int(np.ceil(np.log2(N)))
+    b = [i for i in range(n+1)]
+    x = [n+1+i for i in range(n)]
+    ancilla = 2*n + 1
+    q_reg = 2*n + 2
+    circuit = Circuit(2*n+3)
+    print(f'  - Total number of qubits used: {2*n+3}.\n')
+    r = []
+    
+    # Building the quantum circuit
+    circuit.add(gates.H(q_reg))
+    circuit.add(gates.X(x[len(x)-1]))
+    a_i = (a**(2**(2*n - 1)))
+    circuit.add(c_U(q_reg, x, b, a_i, N, ancilla, n))
+    circuit.add(gates.H(q_reg))
+    circuit.add(gates.M(q_reg))
+    result = circuit(nshots=1)
+    r.append(result.frequencies(binary=False).most_common()[0][0])
+    initial_state = circuit.final_state
+    
+    # Using multiple measurements for the semiclassical QFT.
+    for i in range(1, 2*n):
+        circuit = Circuit(2*n+3)
+        circuit.add(gates.Collapse(q_reg, r[-1]))
```

Comment:
I believe the collapse gate here does not work as intended and shoud be changed as:
```suggestion
        circuit.add(gates.Collapse(q_reg, result=[r[-1]]))
```

The reason is that the collapse gate definition is `gates.Collapse(*q, result=0)` which means that it accepts multiple qubits as targets. Therefore when you do `Collapse(q_reg, r[-1])` you create a collapse gate acting on qubits `(q_reg, r[-1])` with `result=0` (the default), while what you really want is to collapse `q_reg` to `result=r[-1]`, right?

There are a few other issues I found with the Collapse gate which I will fix in another PR. For example, here ideally you should be able to do `result=r[-1]` (without the list `[r[-1]]`), however if you try this now it will fail (I will fix this). The `result=` keyword should definitely be used when you pass a result, because otherwise he will treat the result number as an additional qubit id.

What I still don't understand is how the algorithm works despite this mistake in the implementation.

Reply:
I see, was it like this when we first implemented Collapse? I think I did not change much from our first implementation. I will fix this. 

If I have to guess why it still seems to work is because the collapse fixes the measured qubit to either 1 or 0, so the ones that coincide with 0 still work. But that could explain why the probability of success seemed lower than I expected.